File: leetcode/39. 组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */

// @lc code=start
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<vector<int>> ans;
    void dfs(vector<int>& candidates, vector<int>& combine, int target, int idx){
        if(idx==candidates.size()){
            return;
        }
        if(target==0){
            ans.emplace_back(combine);
            return;
        }
        //使用当前数字
        if(target-candidates[idx]>=0){
            combine.emplace_back(candidates[idx]);
            dfs(candidates, combine, target-candidates[idx], idx);
            combine.pop_back();
        }
        //使用下一个数字
        if(idx+1 < candidates.size() && target-candidates[idx+1]>=0){
            dfs(candidates, combine, target, idx+1);
        }        
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int>combine;
        sort(candidates.begin(),candidates.end());
        dfs(candidates, combine, target, 0);
        return ans;
    }
};
// @lc code=end

File: leetcode/39. 组合总和.py
# 给你一个 无重复元素 的整数数组 candidates 和一个目标整数 target ，找出 candidates 中可以使数字和为目标数 target 的 所有 不同组合 ，并以列表形式返回。你可以按 任意顺序 返回这些组合。

# candidates 中的 同一个 数字可以 无限制重复被选取 。如果至少一个数字的被选数量不同，则两种组合是不同的。 

# 对于给定的输入，保证和为 target 的不同组合数少于 150 个。

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:

        res=[]
        def dfs(path, sum, start):

            if sum==target:
                res.append(path)
                return
            
            if sum>target:
                return
            
            for i in range(start, len(candidates)):
                dfs(path+[candidates[i]], sum+candidates[i], i)

        dfs([],0,0)
        return res
